File: Assets/src/model/enemies/Bloodstone.cs
using System;
using System.Linq;
using System.Runtime.Serialization;
using UnityEngine;

[Serializable]
[ObjectInfo("bloodstone", flavorText: "", description: "You deal +1 attack damage.\nYou take +1 attack damage.\n\nDestroy the Bloodstone to remove.")]
public class Bloodstone : Body {
  BloodstoneStatus status;

  public Bloodstone(Vector2Int pos) : base(pos) {
    hp = baseMaxHp = 1;
    status = new BloodstoneStatus(this);
  }

  protected override void HandleEnterFloor() {
    base.HandleEnterFloor();
    if (floor == GameModel.main.player?.floor) {
      GameModel.main.player.statuses.Add(status);
    }
    RegisterEntityAddedListener();
  }

  [OnDeserialized]
  public void RegisterEntityAddedListener() {
    floor.OnEntityAdded += HandleEntityAdded;
  }

  protected override void HandleLeaveFloor() {
    floor.OnEntityAdded -= HandleEntityAdded;
    base.HandleLeaveFloor();
    status.Refresh();
  }

  private void HandleEntityAdded(Entity entity) {
    if (entity is Player p) {
      p.statuses.Add(status);
    }
  }
}

[Serializable]
[ObjectInfo("bloodstone")]
public class BloodstoneStatus : Status, IAttackDamageModifier, IAttackDamageTakenModifier, IFloorChangeHandler {
  private Bloodstone owner;

  public BloodstoneStatus(Bloodstone owner) {
    this.owner = owner;
  }

  public override bool Consume(Status otherParam) {
    return false;
  }

  public void Refresh() {
    var isOwnerOnDifferentFloor = actor.floor != owner.floor;
    var isOwnerDead = owner.IsDead;
    if (isOwnerOnDifferentFloor || isOwnerDead) {
      Remove();
    }
  }

  public override void HandleFloorChanged(Floor newFloor, Floor oldFloor) {
    if (newFloor != null) {
      Refresh();
    }
  }

  public override string Info() => ObjectInfo.GetDescriptionFor(owner);

  // this is super hacky but this implementation is executed for both IAttackDamage and IAttackDamageTaken
  public int Modify(int input) {
    return input + 1;
  }
}

File: web/src/model/enemies/Bloodstone.ts
import { Body } from '../Body';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { BloodstoneStatus } from '../statuses/BloodstoneStatus';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Entity } from '../Entity';
import type { Player } from '../Player';

/**
 * Applies BloodstoneStatus to the player: +1 attack damage dealt and taken.
 * Destroy the Bloodstone to remove the effect.
 * Port of C# Bloodstone.cs — note: extends Body, not AIActor.
 */
export class Bloodstone extends Body {
  private status: BloodstoneStatus;

  constructor(pos: Vector2Int) {
    super(pos);
    this._hp = this._baseMaxHp = 1;
    this.status = new BloodstoneStatus(this);
  }

  protected handleEnterFloor(): void {
    super.handleEnterFloor();
    const player = GameModelRef.mainOrNull?.player;
    if (player && player.floor === this.floor) {
      player.statuses.add(this.status);
    }
    this.floor!.onEntityAdded.on(this.handleEntityAdded);
  }

  protected handleLeaveFloor(): void {
    this.floor!.onEntityAdded.off(this.handleEntityAdded);
    super.handleLeaveFloor();
    this.status.refresh();
  }

  private handleEntityAdded = (entity: Entity): void => {
    // Check if a Player entered this floor
    if (entity.constructor.name === 'Player' && 'statuses' in entity) {
      (entity as unknown as Player).statuses.add(this.status);
    }
  };
}

entityRegistry.register('Bloodstone', Bloodstone);
